### ISSUE TITLE
Persistent Storage

### ISSUE THREAD

**Issue by kwmc-lucas** [rednafi/fly-fastapi#1]
Hey, I love this example and I've been exploring exactly the same thing!

Something that has me confused, is whether the data will be kept across reboots and instances.

For example, if you have a database receiving POSTs and PUTs, what happens if another instance is spun up in another region, Fly decide to swap VMs, or it needs to be rebooted - won't it be pulling from the raw image/container and therefore lose all of the changes/data?

I've noticed _[Persistent Storage](https://fly.io/docs/about/pricing/#persistent-storage-volumes)_ mentioned, but I'm not sure if it that would solve things and I'm struggling to point FastAPI's database to a mounted volume on Fly. Have you noticed any discrepancies with data?

**Comment by rednafi** [comment]
Hey @kwmc-lucas. Nope. There's actually no data involved in this example. So, nothing to persist here. You're right. If there were any data involved, that wouldn't have survived a reboot. 

To answer your other question, if you have a dedicated database like PostgreSQL, that wouldn't be affected by a VM reboot. If you have 3 application containers and 1 database instance, adding or removing the app containers won't lose any data. If you had only one container, even in that case, rebooting wouldn't lose any data; you just couldn't post anything while the instance boots up. Does this make sense?

**Comment by rednafi** [comment]
There's one interesting idea that I'm meaning to explore. That is—using the free persistent storage to save SQLite files and build apps off of that. This will make sure that the data survives a hard reboot.

**Comment by kwmc-lucas** [comment]
Thanks so much for the quick response.

So I should look into linking FastAPI with a PostgresSQL database for persistency? If it were a single VM on fly.io, new stuff POSTed to the database after deployment (and not in the original image) would survive reboots?

That is interesting. So if you did it that way, there'd be one persistent DB and then apps would pull/push from that? I think that's a cool idea. I guess it would be ideal to sync, so then people in a certain region have the data close to them too?

**Comment by rednafi** [comment]
Yeah if you're building something with high throughput, then take a look at their PostgreSQL docs. They have examples with Go and NodeJS to set it up and connect to the DB. It shouldn't be too hard to hook up a Python application with PostgreSQL. Another great thing about this approach is that you can connect multiple apps to the same DB and separate out the data in different PostgreSQL schemas. 

As for the second approach, you can create a persistent volume with a single command and save your SQLite database there. Then your apps can pull and push data in that and it'll survive any restart whatsoever. I'll experiment with it and add a POST request to the examples in this repo.

**Comment by kwmc-lucas** [comment]
That's great, thank you for that and I also really appreciate you sharing this work.

**Comment by rednafi** [comment]
Yeah, the docs aren't all-encompassing. Thanks for sharing the link.
